solver/src/encoding.rs

```diff
@@ -37,6 +39,51 @@ fn encode_order_flags(order: &UserOrder) -> u8 {
     result
 }
 
+fn encode_uniswap_call(
+    amount_in: U256,
+    amount_out_min: U256,
+    token_in: H160,
+    token_out: H160,
+    payout_to: H160,
+) -> [u8; UNISWAP_INTERACTION_SIZE] {
```

Comment:
Wouldn't it be more convenient if this method returned Bytes (or Vec<u8>). Then our code wouldn't have to know the stride of the encoded call which seems more flexible. Also with multiple interactions we likely have to do a lot of concatenation and eventually things will be submitted as Bytes to the smart contract.

Reply:
The caller can always call to_vec. I'm not very opposed to making this return a vec directly just conceptually this has a fixed size so that's what this low level method should return.

Reply:
I also prefer it as `[u8; UNISWAP_INTERACTION_SIZE]`. The  caller can always `Vec::extend_from_slice` to build many of these or even `.to_vec()` to convert it to a vector. My main argument is that the using an array instead of a vec encodes the length information in the type instead of just losing it.

Reply:
Sure, but why do we care about the length? If the Uniswap signature changed (which is impossible but let's assume this was another "upgradable" interaction), we would have to change the size constant as well. And right now it wouldn't even give us a compilation error but likely just panic at runtime when it tries to `copy_from_slice` from a differently sized slice. 

My point is that I don't see what part of the solver or even contract for that matter actually cares about how many bytes an interaction has - isn't it just an arbitrarily sized byte vector?

Reply:
> Sure, but why do we care about the length?

It just enriches the type with more information.

> If the Uniswap signature changed (which is impossible but let's assume this was another "upgradable" interaction), we would have to change the size constant as well.

The idea is that then the implementation of `encode_uniswap_call` would be incorrect, but not the outside code that compiles with the `[u8; UNISWAP_INTERACTION_SIZE]` type. Then, if the size would change, the compiler would help catch errors that would cause panics at runtime (like indexing out of bounds):

```rust
    let array = [0u8; 20];
    println!("{}", array[21]);
```

Produces an error
```
error: this operation will panic at runtime
 --> src/main.rs:3:20
  |
3 |     println!("{}", array[21]);
  |                    ^^^^^^^^^ index out of bounds: the length is 20 but the index is 21
```

By giving more information to the compiler just enables it to do more checking. This would not be the case with `Vec`.

Additionally, calling code would have to make less assumptions about the length of the encoded bytes (for example, does indexing the `encode_uniswap_call` result need to be checked everytime for errors?) and in unit tests in calling code passing a `[u8; UNISWAP_INTERACTION_SIZE]` make it impossible to create some bytes with the wrong length, this means if the size of `UNISWAP_INTERACTION_SIZE` would change because it is an upgradable contract, then these tests would have to manually be found and adjusted and the compiler would not be able to help, making it possible for calling code to have bugs that won't get caught by the unit tests and cause runtime errors.

> just panic at runtime when it tries to copy_from_slice from a differently sized slice.

It would break the `encode_uniswap_call` unit test.

Reply:
> My point is that I don't see what part of the solver or even contract for that matter actually cares about how many bytes an interaction has - isn't it just an arbitrarily sized byte vector?

In that case:
```rust
type UniswapInteraction = [u8; UNISWAP_INTERACTION_SIZE];

trait ContractInteraction: AsRef<[u8]> {}
impl ContractInteraction for UniswapInteraction {}
```

Reply:
> Then, if the size would change, the compiler would help catch errors that would cause panics at runtime (like indexing out of bounds)

Believe me that I generally love type safety. My point is given that we treat contract interactions as arbitrary bytes, they **should** not be indexed into or otherwise inspected by the caller (they should only concatenate them to larger calldata). Your point about the unit test failing and thus catching potential changes is good. 

I don't think this debate is blocking the PR, feel free to merge as is.

Reply:
I'm also fine with using a `Vec` for the return type, it's really a very small detail that I don't think it's important. I just like using a fixed size array slightly better as I feel the type more accurately describes what is being returned (a Uniswap contract interaction vs a general contract interaction) and it can always be converted to a `Vec` with `to_vec`. This allows the caller to essentially decide whether or not it wants the type erasure.